usr/opt/OSC200/usr/share/man/man3/rexec.3
.\" This manpage source uses rsml coding. 
.so /usr/share/lib/tmac/sml
.so /usr/share/lib/tmac/rsml
.\"
.\"
.\" *********************************************************************
.\" *                                                                   *
.\" *      Copyright (c) Digital Equipment Corporation, 1991, 1994      *
.\" *                                                                   *
.\" *                       All Rights Reserved.                        *
.\" *                                                                   *
.\" *********************************************************************
.\"
.\" HISTORY
.\"
.\" @(#)$RCSfile: rexec.3,v $ $Revision: 3.2.5.4 $ (DEC) $Date: 1994/01/20 02:06:19 $
.\" 
.\" (c) Copyright 1991, OPEN SOFTWARE FOUNDATION, INC.
.\" ALL RIGHTS RESERVED
.\" 
.TH rexec 3
.SH NAME
.PP
\*Lrexec\*O, \*Lrexec_r\*O \- Allow command execution on a remote host
.SH LIBRARY
.PP
Standard C Library \*L(libc.a)
.SH SYNOPSIS
.PP
.sS
\*L#include <netdb.h>
.sp .4v
\*Lint rexec\*O(\*L
.nL
	\*Lchar **\*Vhost\*L,
.nL
	\*Lint \*Vport\*L,
.nL
	char *\*Vuser\*L,
.nL
	char *\*Vpasswd\*L,
.nL
	char *\*Vcommand\*L,
.nL
	\*Lint *\*Verr_file_desc\*L);
.nL
\*Lint rexec_r\*O(\*L
.nL
	\*Lchar **\*Vhost\*L,
.nL
	\*Lint \*Vport\*L,
.nL
	char *\*Vuser\*L,
.nL
	char *\*Vpasswd\*L,
.nL
	char *\*Vcommand\*L,
.nL
	\*Lint *\*Verr_file_desc\*L,
.nL
	\*Lstruct hostent_data \*V*host_data\*O);
.sE
.SH PARAMETERS
.PP
.VL
.LI "\*Vhost"
\*OContains
the name of a remote host that is listed in the \*L/etc/hosts\*O
file or \*L/etc/resolv.conf
\*Ofile.  If the name of the host is not found in either file,
the \*Lrexec(\|)\*O fails.
.LI "\*Vport"
\*OSpecifies
the \*Owell-known Internet
port to use for the connection.  A pointer to the structure that contains
the necessary port can be obtained by issuing the following library
call:
.oS
getservbyname("exec", "tcp")
.oE
.LI "\*Vuser\*O"
\*OPoints to a user ID that is valid on the host.
.LI "\*Vpasswd"
\*OPoints to the password of the specified user ID on the host.
.LI "\*Vcommand"
\*OPoints to
the name of the command to be executed at the remote host.
.LI "\*Verr_file_desc"
\*OSpecifies the file to which standard error from the remote command
is sent.
.sp
If the \*Verr_file_desc\*O parameter is 0 (zero),
the standard error of the remote command is the same
as standard output.  No provision is made for sending arbitrary
signals to the remote process.  In this case, however, it may be possible
to send \*Oout-of-band data to the
remote command.
.sp
If the \*Verr_file_desc\*O parameter is nonzero, an auxiliary channel to a control process is set
up, and a descriptor for it is placed in the \*Verr_file_desc \*Oparameter.
The control process provides diagnostic output from the remote
command on this channel and also accepts bytes as signal numbers to
be forwarded to the process group of the command.  This diagnostic
information does not include remote authorization failure, since this
connection is set up after authorization has been verified.
.LI "\*Vhost_data\*O"
Per-thread data for reentrant version.
.LE
.SH DESCRIPTION
.PP
.iX "rexec function"
.iX "remote host" "executing commands on"
.iX "commands" "executing on remote host"
.iX "executing commands on remote host"
The \*Lrexec(\|)\*O (remote execution) function allows the calling
process to execute commands on a remote host.
.PP
If the \*Lrexec(\|)\*O connection succeeds, a socket in the Internet
domain of type
\*LSOCK_STREAM\*O is returned
to the calling process and is given to the remote command as standard
input and standard output.
.PP
The
\*Vuser\*O
and
\*Vpasswd\*O
parameters specify a valid user ID and the associated
password for that user on the remote host.
If the \*Vuser\*O and
\*Vpasswd\*O
parameters are not supplied, the \*Lrexec(\|)\*O function takes the
following actions until finding a user ID and password to send to
the remote host:
.AL
.LI
\*OIt searches the current environment for the user ID and password
on the remote host.
.LI
\*OIt searches the user's home directory for a file called \*L$HOME/.netrc\*O
that contains
a user ID and password.
.LI
\*OIt prompts the user for a user ID and password.
.LE
.PP
Note that action 3 does not apply when using
\*Lrexec_r(\|)\*O.
.PP
The \*Lrexec_r(\|)\*O function is the reentrant version of
\*Lrexec(\|)\*O.  The \*Lnetdb.h\*O header file defines the
\*Lhostent_data\*O structures.
.SH RETURN VALUES
.PP
Upon successful completion, the system returns a socket
descriptor
to the remote command.
\*OOtherwise, a value of
-1 is returned, indicating that the specified hostname does not
exist.
.SH FILES
.PP
.VL
.LI "\*L/etc/hosts"
\*OContains
hostnames and their addresses for hosts in a network.  This file
is used to resolve a hostname into an Internet address.
.LI "\*L/etc/resolv.conf"
\*OContains the name server and domain
name.
.LI "\*L$HOME/.netrc"
\*OContains
automatic login information.
.LE
.SH RELATED INFORMATION
.PP
Functions:
\*Lgetservbyname\*O(3),
\*Lrcmd\*O(3),
\*Lrresvport\*O(3),
\*Lruserok\*O(3).
.PP
Commands:
\*Lrexecd\*O(8).
.iX "rexec function"
.EQ
delim off
.EN


usr/opt/OSC200/src/usr/ccs/lib/libc/rexec.c
/*
 * *****************************************************************
 * *                                                               *
 * *    Copyright (c) Digital Equipment Corporation, 1991, 1994    *
 * *                                                               *
 * *   All Rights Reserved.  Unpublished rights  reserved  under   *
 * *   the copyright laws of the United States.                    *
 * *                                                               *
 * *   The software contained on this media  is  proprietary  to   *
 * *   and  embodies  the  confidential  technology  of  Digital   *
 * *   Equipment Corporation.  Possession, use,  duplication  or   *
 * *   dissemination of the software and media is authorized only  *
 * *   pursuant to a valid written license from Digital Equipment  *
 * *   Corporation.                                                *
 * *                                                               *
 * *   RESTRICTED RIGHTS LEGEND   Use, duplication, or disclosure  *
 * *   by the U.S. Government is subject to restrictions  as  set  *
 * *   forth in Subparagraph (c)(1)(ii)  of  DFARS  252.227-7013,  *
 * *   or  in  FAR 52.227-19, as applicable.                       *
 * *                                                               *
 * *****************************************************************
 */
/*
 * HISTORY
 */
#ifndef lint
static char	*sccsid = "@(#)$RCSfile: rexec.c,v $ $Revision: 4.2.8.2 $ (DEC) $Date: 1993/06/07 22:27:52 $";
#endif 
/*
 * (c) Copyright 1990, 1991, 1992, 1993 OPEN SOFTWARE FOUNDATION, INC. 
 * ALL RIGHTS RESERVED 
 */
/*
 * OSF/1 1.2
 */
#if !defined(lint) && !defined(_NOIDENT)
#endif
/*
 * Copyright International Business Machines Corp. 1989
 * Unpublished Work
 * All Rights Reserved
 * Licensed Material - Property of IBM
 */
/*
 * Copyright (c) 1980 Regents of the University of California.
 * All rights reserved.  The Berkeley software License Agreement
 * specifies the terms and conditions for redistribution.
 *
 * rexec.c	1.2  com/lib/c/net,3.1,8943 7/21/89 08:08:08
 */

/* name space pollution clean up */
#ifdef _NAME_SPACE_WEAK_STRONG
#include "pollution.h"
#if defined(_THREAD_SAFE)
#pragma weak rexec_r = __rexec_r
#endif
#if !defined(_THREAD_SAFE)
#pragma weak rexec = __rexec
#endif
#endif
#include <sys/param.h>
#include <sys/socket.h>
#include <netinet/in.h>

#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <netdb.h>
#include <errno.h>
#include <nl_types.h>
#include "libc_msg.h"

static nl_catd catd;

int
#ifdef _THREAD_SAFE
rexec_r(char **ahost, int rport, char *name, char *pass, char *cmd, int *fd2p,
	struct hostent_data *host_data)
#else
rexec(char **ahost, int rport, char *name, char *pass, char *cmd, int *fd2p)
#endif	/* _THREAD_SAFE */
{
	int s, timo = 1, s3;
	struct sockaddr_in sin, sin2, from;
	char c;
	short port=0;
	struct hostent *hp;

#ifdef	_THREAD_SAFE
	struct hostent host;
	hp = &host;

	if (gethostbyname_r(*ahost, hp, host_data) < 0)
#else
	hp = gethostbyname(*ahost);
	if (hp == 0)
#endif
	{
                catd = catopen(MF_LIBC,0);
                fprintf(stderr, catgets(catd,LIBCNET,NET22,
                                        "%s: unknown host\n"), *ahost);
		catclose(catd);
		return (-1);
	}
	*ahost = hp->h_name;
	ruserpass(hp->h_name, &name, &pass);
retry:
	s = socket(AF_INET, SOCK_STREAM, 0);
	if (s < 0) {
		perror("rexec: socket");
		return (-1);
	}
	sin.sin_family = hp->h_addrtype;
	sin.sin_port = rport;
	bcopy(hp->h_addr, (caddr_t)&sin.sin_addr, hp->h_length);
	if (connect(s, (struct sockaddr *)&sin, sizeof(sin)) < 0) {
		if (_Geterrno() == ECONNREFUSED && timo <= 16) {
			(void) close(s);
			sleep(timo);
			timo *= 2;
			goto retry;
		}
		perror(hp->h_name);
		return (-1);
	}
	if (fd2p == 0) {
		(void) write(s, "", 1);
		port = 0;
	} else {
		char num[8];
		int s2, sin2len;
		
		s2 = socket(AF_INET, SOCK_STREAM, 0);
		if (s2 < 0) {
			(void) close(s);
			return (-1);
		}
		listen(s2, 1);
		sin2len = sizeof (sin2);
		if (getsockname(s2, (struct sockaddr *)&sin2, &sin2len) < 0 ||
		  sin2len != sizeof (sin2)) {
			perror("getsockname");
			(void) close(s2);
			goto bad;
		}
		port = ntohs((u_short)sin2.sin_port);
		(void) sprintf(num, "%d", port);
		(void) write(s, num, strlen(num)+1);
		{ int len = sizeof (from);
		  s3 = accept(s2, (struct sockaddr *)&from, &len);
		  close(s2);
		  if (s3 < 0) {
			perror("accept");
			port = 0;
			goto bad;
		  }
		}
		*fd2p = s3;
	}
	(void) write(s, name, strlen(name) + 1);
	/* should public key encypt the password here */
	(void) write(s, pass, strlen(pass) + 1);
	(void) write(s, cmd, strlen(cmd) + 1);
	if (read(s, &c, 1) != 1) {
		perror(*ahost);
		goto bad;
	}
	if (c != 0) {
		while (read(s, &c, 1) == 1) {
			(void) write(2, &c, 1);
			if (c == '\n')
				break;
		}
		goto bad;
	}
	return (s);
bad:
	if (port)
		(void) close(*fd2p);
	(void) close(s);
	return (-1);
}
